In [ ]:
#Variavel com RU
ru = 3727967

In [ ]:
# importar o arquivo diretamente de seu google drive
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
#A primeira linha instala o Java 8 na maquina da sessão
#A segunda faz o download do spark no link escolhido
#A terceira extrai o arquivo baixado

!apt-get install openjdk-8-jdk-headless -qq > /dev/null
!wget -q https://dlcdn.apache.org/spark/spark-3.4.0/spark-3.4.0-bin-hadoop3.tgz
!tar xf spark-3.4.0-bin-hadoop3.tgz

In [ ]:
#E a quarta instala a biblioteca findspark para facilitar na criação da sessão spark

!pip install -q findspark

In [ ]:
#Agora iremos utilizar o código abaixo para configurar as variáveis de caminho de ambiente utilizando a biblioteca "os"

import os

os.environ["JAVA_HOME"] = "/usr/lib/jvm/java-8-openjdk-amd64"
os.environ["SPARK_HOME"] = "/content/spark-3.4.0-bin-hadoop3"

In [ ]:
#criando uma copia do dataframe na minha pasta drive
!cp /content/drive/MyDrive/Colab_Notebooks/imdb-reviews-pt-br.csv /content

In [ ]:
#importando e iniciando o findspark e importando o sparksession
import findspark

findspark.init()

from pyspark.sql import SparkSession

In [ ]:
#criando a sessao
spark = SparkSession.builder.master("local[*]").getOrCreate()

# **Prática 01 – Somatório de IDs. RU - 3727967**
**Qual o valor da soma de todos os campos “id” dos filmes classificados como negativos para o banco de dados “imdb-reviews-pt-br.csv”**


In [ ]:
#carregando o conjunto de dados
df = spark.read.csv("imdb-reviews-pt-br.csv", header=True,quote="\"", escape="\"", encoding="UTF-8")

In [ ]:
#mostrando as 20 primeiras linhas
df.show()

+---+--------------------+--------------------+---------+
| id|             text_en|             text_pt|sentiment|
+---+--------------------+--------------------+---------+
|  1|Once again Mr. Co...|Mais uma vez, o S...|      neg|
|  2|This is an exampl...|Este é um exemplo...|      neg|
|  3|First of all I ha...|Primeiro de tudo ...|      neg|
|  4|Not even the Beat...|Nem mesmo os Beat...|      neg|
|  5|Brass pictures mo...|Filmes de fotos d...|      neg|
|  6|A funny thing hap...|Uma coisa engraça...|      neg|
|  7|This German horro...|Este filme de ter...|      neg|
|  8|Being a long-time...|Sendo um fã de lo...|      neg|
|  9|"Tokyo Eyes" tell...|"Tokyo Eyes" fala...|      neg|
| 10|Wealthy horse ran...|Fazendeiros ricos...|      neg|
| 11|Cage plays a drun...|Cage interpreta u...|      neg|
| 12|First of all, I w...|Primeiro de tudo,...|      neg|
| 13|So tell me - what...|Então me diga - q...|      neg|
| 14|A big disappointm...|Uma grande decepç...|      neg|
| 15|This film

In [ ]:
#print do schema
df.printSchema()

root
 |-- id: string (nullable = true)
 |-- text_en: string (nullable = true)
 |-- text_pt: string (nullable = true)
 |-- sentiment: string (nullable = true)



In [ ]:
#confirmando o tipo do dataframe
type(df)

pyspark.sql.dataframe.DataFrame

In [ ]:
#contando  as linhas do conjunto de dados
df.count()

49459

In [ ]:
# importando os métodos com funções para transformações de variáveis
from pyspark.sql.functions import *

In [ ]:
#Transformando a coluna id em integer.
print(f"Meu RU: {ru}")
df = df.withColumn(
    colName = 'id', 
    col = regexp_replace('id','','').cast('int')
    )

# printando o resultado
df.select('id').printSchema()

Meu RU: 3727967
root
 |-- id: integer (nullable = true)



In [ ]:
#Criando variavel apenas com coluna id e sentiment
df_Select = df.select('id','sentiment')

In [ ]:
#Mostrando o resultado
print(f"Meu RU: {ru}")
df_Select.show()

Meu RU: 3727967
+---+---------+
| id|sentiment|
+---+---------+
|  1|      neg|
|  2|      neg|
|  3|      neg|
|  4|      neg|
|  5|      neg|
|  6|      neg|
|  7|      neg|
|  8|      neg|
|  9|      neg|
| 10|      neg|
| 11|      neg|
| 12|      neg|
| 13|      neg|
| 14|      neg|
| 15|      neg|
| 16|      neg|
| 17|      neg|
| 18|      neg|
| 19|      neg|
| 20|      neg|
+---+---------+
only showing top 20 rows



In [ ]:
#Filtrando por sentiment = neg
df_Neg = df_Select.where('sentiment == "neg" ')


In [ ]:
#Mostrando resultado
print(f"Meu RU: {ru}")
df_Neg.show()

Meu RU: 3727967
+---+---------+
| id|sentiment|
+---+---------+
|  1|      neg|
|  2|      neg|
|  3|      neg|
|  4|      neg|
|  5|      neg|
|  6|      neg|
|  7|      neg|
|  8|      neg|
|  9|      neg|
| 10|      neg|
| 11|      neg|
| 12|      neg|
| 13|      neg|
| 14|      neg|
| 15|      neg|
| 16|      neg|
| 17|      neg|
| 18|      neg|
| 19|      neg|
| 20|      neg|
+---+---------+
only showing top 20 rows



In [ ]:
#somando a coluna id com base no sentiment neg.
print(f"Meu RU: {ru}")
total_neg = df_Neg.groupBy('sentiment').agg(sum('id').alias("Total_Id_Neg")).orderBy('sentiment').where('sentiment == "neg" ')
total_neg.show()

Meu RU: 3727967
+---------+------------+
|sentiment|Total_Id_Neg|
+---------+------------+
|      neg|   459568555|
+---------+------------+



In [ ]:
#Pegando o valor da coluna Total_Id_Neg e printando resultado
soma_neg = total_neg.select(col("Total_Id_Neg")).first()[0]
print(f'Meu RU: {ru}. Valor do somatorio dos ids negativos: {soma_neg}')

Meu RU: 3727967. Valor do somatorio dos ids negativos: 459568555


# **O valor da soma do id com base no sentiment = neg é: 459568555**

---

# **Prática 02 – Diferença do número de palavras totais de português para inglês dos textos negativos RU - 3727967**

**Contar palavras dos textos negativos e achar diferença de quantidade.**

In [ ]:
#selecionando a coluna desejada
df_en = df.select('id','text_en','sentiment')
df_pt = df.select('id','text_pt','sentiment')

In [ ]:
#filtrando o conjunto de dados para o sentiment = neg
df_en = df_en.where('sentiment == "neg" ')
df_pt = df_pt.where('sentiment == "neg" ')

In [ ]:
#removendo a coluna sentiment
df_en = df_en.drop('sentiment')
df_pt = df_pt.drop('sentiment')

In [ ]:
print(f"Meu RU: {ru}")
df_en.show()
df_pt.show()

Meu RU: 3727967
+---+--------------------+
| id|             text_en|
+---+--------------------+
|  1|Once again Mr. Co...|
|  2|This is an exampl...|
|  3|First of all I ha...|
|  4|Not even the Beat...|
|  5|Brass pictures mo...|
|  6|A funny thing hap...|
|  7|This German horro...|
|  8|Being a long-time...|
|  9|"Tokyo Eyes" tell...|
| 10|Wealthy horse ran...|
| 11|Cage plays a drun...|
| 12|First of all, I w...|
| 13|So tell me - what...|
| 14|A big disappointm...|
| 15|This film is abso...|
| 16|Heres a decidedly...|
| 17|At the bottom end...|
| 18|Earth has been de...|
| 19|Many people are s...|
| 20|New York family i...|
+---+--------------------+
only showing top 20 rows

+---+--------------------+
| id|             text_pt|
+---+--------------------+
|  1|Mais uma vez, o S...|
|  2|Este é um exemplo...|
|  3|Primeiro de tudo ...|
|  4|Nem mesmo os Beat...|
|  5|Filmes de fotos d...|
|  6|Uma coisa engraça...|
|  7|Este filme de ter...|
|  8|Sendo um fã de lo...|
|  9|"Tokyo E

In [ ]:
# Usa as funções explode e split para dividir o texto em palavras
palavras_pt = df_pt.select("id", explode(split("text_pt", "[ ]+")).alias("palavra"))
palavras_en = df_en.select("id", explode(split("text_en", "[ ]+")).alias("palavra"))

# Usa a função count para contar as ocorrências de cada palavra ordenando do maior para o menor
contagem_pt = palavras_pt.groupBy("palavra").count().orderBy(desc("count"))
contagem_en = palavras_en.groupBy("palavra").count().orderBy(desc("count"))

# Mostra o resultado
contagem_pt.show()
contagem_en.show()

+-------+------+
|palavra| count|
+-------+------+
|     de|209113|
|    que|167678|
|      e|137200|
|      o|123049|
|     um|107492|
|      a|102794|
|      é| 93327|
|    não| 73713|
|    uma| 64633|
|   para| 64105|
|     em| 63829|
|  filme| 61801|
|     do| 49849|
|    com| 48684|
|     eu| 37822|
|     se| 37108|
|   como| 36659|
|     os| 36428|
|    mas| 33143|
|    por| 31171|
+-------+------+
only showing top 20 rows

+-------+------+
|palavra| count|
+-------+------+
|    the|265718|
|      a|145812|
|    and|131834|
|     of|128764|
|     to|127790|
|     is| 91404|
|     in| 76328|
|      I| 69703|
|   that| 61670|
|   this| 60545|
|     it| 53013|
|    was| 48979|
|    for| 38483|
|   with| 37810|
|     as| 35494|
|    The| 34522|
|  movie| 33879|
|    but| 32892|
|     on| 29428|
|   have| 28906|
+-------+------+
only showing top 20 rows



In [ ]:
# Usa as funções explode e split para dividir o texto em palavras
palavras_pt = df_pt.select("id", explode(split("text_pt", "[ ]+")).alias("palavra"))
palavras_en = df_en.select("id", explode(split("text_en", "[ ]+")).alias("palavra"))

# Usa a função count para contar as ocorrências de cada palavra ordenando do maior para o menor
contagem_pt = palavras_pt.groupBy("palavra").count().orderBy(desc("count"))
contagem_en = palavras_en.groupBy("palavra").count().orderBy(desc("count"))

# Mostra o resultado
contagem_pt.show()
contagem_en.show()

+-------+------+
|palavra| count|
+-------+------+
|     de|209113|
|    que|167678|
|      e|137200|
|      o|123049|
|     um|107492|
|      a|102794|
|      é| 93327|
|    não| 73713|
|    uma| 64633|
|   para| 64105|
|     em| 63829|
|  filme| 61801|
|     do| 49849|
|    com| 48684|
|     eu| 37822|
|     se| 37108|
|   como| 36659|
|     os| 36428|
|    mas| 33143|
|    por| 31171|
+-------+------+
only showing top 20 rows

+-------+------+
|palavra| count|
+-------+------+
|    the|265718|
|      a|145812|
|    and|131834|
|     of|128764|
|     to|127790|
|     is| 91404|
|     in| 76328|
|      I| 69703|
|   that| 61670|
|   this| 60545|
|     it| 53013|
|    was| 48979|
|    for| 38483|
|   with| 37810|
|     as| 35494|
|    The| 34522|
|  movie| 33879|
|    but| 32892|
|     on| 29428|
|   have| 28906|
+-------+------+
only showing top 20 rows



In [ ]:
#Soma dos valores
contagem_pt.agg(sum('count').alias("Total_palavras_pt")).show()
contagem_en.agg(sum('count').alias("Total_palavras_en")).show()

+-----------------+
|Total_palavras_pt|
+-----------------+
|          5455273|
+-----------------+

+-----------------+
|Total_palavras_en|
+-----------------+
|          5400276|
+-----------------+



In [ ]:
#pegando os valores para fazer a subtração
tot_pt = contagem_pt.agg(sum('count').alias("Total_palavras_pt")).first()[0]   
tot_en = contagem_en.agg(sum('count').alias("Total_palavras_en")).first()[0]   

In [ ]:
tot_pt, tot_en

(5455273, 5400276)

In [ ]:
print(f'Meu RU: {ru}". O texto em Portugues possui {tot_pt - tot_en} palavras a mais que o ingles')

Meu RU: 3727967". O texto em Portugues possui 54997 palavras a mais que o ingles


# **O texto em Portugues possui 54997 palavras a mais que o ingles**